spark/src/main/scala/org/apache/spark/sql/delta/redirect/TableRedirect.scala

```diff
@@ -0,0 +1,311 @@
+ *                                                   redirection by setting the table to
+ *                                                   'DROP-REDIRECT-IN-PROGRESS.' This ensures that
+ *                                                   concurrent DML/DDL operations do not interfere
+ *                                                   with the cancellation process.
+ *   3. DROP-REDIRECT-IN-PROGRESS -> NO-REDIRECT: Completes the removal of table redirection. As a
+ *                                                result, all DML, DDL, and read-only queries are no
+ *                                                longer redirected to the previous destination.
+ *   4. ENABLE-REDIRECT-IN-PROGRESS -> NO-REDIRECT: This transition involves canceling table
+ *                                                  redirection while it is still in the process of
+ *                                                  being enabled.
+ */
+sealed trait RedirectState {
+  val name: String
+}
+
+/** This state indicates that redirect is not enabled on the table. */
+case object NoRedirect extends RedirectState {
+  override val name = "NO-REDIRECT"
+}
+
+/** This state indicates that the redirect process is still going on. */
+case object EnableRedirectInProgress extends RedirectState {
+  override val name = "ENABLE-REDIRECT-IN-PROGRESS"
+}
+
+/**
+ * This state indicates that the redirect process is completed. All types of queries would be
+ * redirected to the table specified inside RedirectSpec object.
+ */
+case object RedirectReady extends RedirectState { override val name = "REDIRECT-READY" }
+
+/**
+ * The table redirection is under withdrawal and the redirection property is going to be removed
+ * from the delta table. In this state, the delta client stops redirecting new queries to redirect
+ * destination tables, and only accepts read-only queries to access the redirect source table.
+ * The on-going redirected write or metadata transactions, which are visiting redirect
+ * destinations, can not commit.
+ */
+case object DropRedirectInProgress extends RedirectState {
+  override val name = "DROP-REDIRECT-IN-PROGRESS"
+}
+
+/**
+ * This is the abstract class of the redirect specification, which stores the information
+ * of accessing the redirect destination table.
+ */
+abstract class RedirectSpec()
+
+/**
+ * The default redirect spec that is used for OSS delta.
+ * This is the specification about how to access the redirect destination table.
+ * One example of its JSON presentation is:
+ *   {
+ *     ......
+ *     "spec": {
+ *       "tablePath": "s3://<bucket-1>/tables/<table-name>"
+ *     }
+ *   }
+ * @param tablePath this is the path where stores the redirect destination table's location.
+ */
+class PathBasedRedirectSpec(val tablePath: String) extends RedirectSpec
+
+object PathBasedRedirectSpec {
+  /**
+   * The default type of redirection is path based redirection. Delta client uses the `tablePath`
```

Comment:
What do you mean by default type? Does this mean we will use `PathBasedRedirect` even if it is not mentioned in the stored spec?

Based on the code - I think thats not the case. we can rephrase the comment.

Reply:
Done